# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: js/src/builtin/FinalizationRegistryObject.h
# Commit: d4be324b80a4
# Full Hash: d4be324b80a403215e60c3462272fbc46241d3cd
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2020-04-10 02:59:09
# Description:
#   Bug 1628440 - Use WeakHeapPtr for weakly-held vector elements in finalization registry registrations weakmap r=sfink
#   
#   The problem is that HeapPtr<> has a prebarrier in its destructor.  This isn't necessary for weakly held values, and cause problems if the values have already been finalized.  The patch also renames FinalizationRecordVectorObject to FinalizationRegistrationsObject to better describe its purpose.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D70346
# ==============================================================================

diff -r ce494e20e11c -r d4be324b80a4 js/src/builtin/FinalizationRegistryObject.h
--- a/js/src/builtin/FinalizationRegistryObject.h	Thu Apr 09 15:34:49 2020 +0000
+++ b/js/src/builtin/FinalizationRegistryObject.h	Thu Apr 09 16:23:12 2020 +0000
@@ -31,7 +31,7 @@
  *   |  |                       v          |        v             |          |
  *   |  |  +--------------------+------+   |   +----+-----+       |          |
  *   |  |  |       Finalization-       |   |   |  Target  |       |          |
- *   |  |  |     RecordVectorObject    |   |   | JSObject |       |          |
+ *   |  |  |    RegistrationsObject    |   |   | JSObject |       |          |
  *   |  |  +-------------+-------------+   |   +----------+       |          |
  *   |  |  |       RecordVector        |   |                      |          |
  *   |  |  +-------------+-------------+   |                      |          |
@@ -54,7 +54,7 @@
  *
  * When a target is registered an unregister token may be supplied. If so, this
  * is also recorded by the registry and is stored in a weak map of
- * registrations. The values of this map are FinalizationRecordVectorObject
+ * registrations. The values of this map are FinalizationRegistrationsObject
  * objects. It's necessary to have another JSObject here because our weak map
  * implementation only supports JS types as values.
  *
@@ -131,24 +131,25 @@
   FinalizationRegistryObject* registryUnbarriered() const;
 };
 
-// A vector of FinalizationRecordObjects.
-using FinalizationRecordVector =
-    GCVector<HeapPtr<FinalizationRecordObject*>, 1, js::ZoneAllocPolicy>;
+// A vector of weakly-held FinalizationRecordObjects.
+using WeakFinalizationRecordVector =
+    GCVector<WeakHeapPtr<FinalizationRecordObject*>, 1, js::ZoneAllocPolicy>;
 
-// A JS object containing a vector of FinalizationRecordObjects, which holds the
-// records corresponding to the registrations for a particular registration
-// token. These are used as the values in the registration weakmap. The contents
-// of the vector are weak references and are not traced.
-class FinalizationRecordVectorObject : public NativeObject {
+// A JS object containing a vector of weakly-held FinalizationRecordObjects,
+// which holds the records corresponding to the registrations for a particular
+// registration token. These are used as the values in the registration
+// weakmap. Since the contents of the vector are weak references they are not
+// traced.
+class FinalizationRegistrationsObject : public NativeObject {
   enum { RecordsSlot = 0, SlotCount };
 
  public:
   static const JSClass class_;
 
-  static FinalizationRecordVectorObject* create(JSContext* cx);
+  static FinalizationRegistrationsObject* create(JSContext* cx);
 
-  FinalizationRecordVector* records();
-  const FinalizationRecordVector* records() const;
+  WeakFinalizationRecordVector* records();
+  const WeakFinalizationRecordVector* records() const;
 
   bool isEmpty() const;
 
@@ -166,6 +167,9 @@
   static void finalize(JSFreeOp* fop, JSObject* obj);
 };
 
+using FinalizationRecordVector =
+    GCVector<HeapPtr<FinalizationRecordObject*>, 1, js::ZoneAllocPolicy>;
+
 using FinalizationRecordSet =
     GCHashSet<HeapPtrObject, MovableCellHasher<HeapPtrObject>, ZoneAllocPolicy>;
 